File: Solutions/2800 - 2899/2807. Insert Greatest Common Divisors in Linked List/Solution.java
class Solution {
    public static int gcd(int a,int b)
    {
        while(b!=0)
        {
            int temp=b;
            b=a%b;
            a=temp;
        }
        return a;
    }
    public ListNode insertGreatestCommonDivisors(ListNode head) {
        if(head==null || head.next==null)
            return head;
        
        ListNode cur = head.next;
        ListNode prev = head;
        
        while(cur!=null)
        {
            int val=gcd(prev.val,cur.val);
            ListNode newNode = new ListNode(val,cur);
            prev.next=newNode;
            prev=cur;
            cur=cur.next;
        }
        return head;
    }
}

File: Solutions/2800 - 2899/2807. Insert Greatest Common Divisors in Linked List/Solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def gcd(a, b):
            while b != 0:
                a, b = b, a%b
            return a

        curr = head
        while curr and curr.next:
            res = gcd(curr.val, curr.next.val)
            new_node = ListNode(res)
            new_node.next = curr.next
            curr.next = new_node
            curr = new_node.next
        
        return head
